src/compiler/checker.ts

```diff
@@ -18022,20 +18029,29 @@ namespace ts {
                 return anyType;
         }
 
-        function checkJsxReturnAssignableToAppropriateBound(isSFC: boolean, elemInstanceType: Type, openingLikeElement: Node) {
-            if (isSFC) {
+        function checkJsxReturnAssignableToAppropriateBound(refKind: JsxReferenceKind, elemInstanceType: Type, openingLikeElement: Node) {
+            if (refKind === JsxReferenceKind.Function) {
                 const sfcReturnConstraint = getJsxStatelessElementTypeAt(openingLikeElement);
                 if (sfcReturnConstraint) {
                     checkTypeRelatedTo(elemInstanceType, sfcReturnConstraint, assignableRelation, openingLikeElement, Diagnostics.JSX_element_type_0_is_not_a_constructor_function_for_JSX_elements);

```

Comment:
(note) as of React 16, components are also allowed to return `number` / `string` / `boolean` (in other words, pretty much anything other than `undefined` or a non-Element, non-Array object), but at least TS@latest will reject SFC which include those return types. It does validate with classes, just not functions.

If fixed this probably goes in another PR though.